src/dash.1
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\" Copyright (c) 1997-2005
.\"	Herbert Xu <herbert@gondor.apana.org.au>.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Kenneth Almquist.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)sh.1	8.6 (Berkeley) 5/4/95
.\"
.Dd January 19, 2003
.Os
.Dt DASH 1
.Sh NAME
.Nm dash
.Nd command interpreter (shell)
.Sh SYNOPSIS
.Nm
.Bk -words
.Op Fl aCefnuvxIimqVEb
.Op Cm +aCefnuvxIimqVEb
.Ek
.Bk -words
.Op Fl o Ar option_name
.Op Cm +o Ar option_name
.Ek
.Bk -words
.Op Ar command_file Oo Ar argument ... Oc
.Ek
.Nm
.Fl c
.Bk -words
.Op Fl aCefnuvxIimqVEb
.Op Cm +aCefnuvxIimqVEb
.Ek
.Bk -words
.Op Fl o Ar option_name
.Op Cm +o Ar option_name
.Ek
.Bk -words
.Ar command_string
.Op Ar command_name Oo Ar argument ... Oc
.Ek
.Nm
.Fl s
.Bk -words
.Op Fl aCefnuvxIimqVEb
.Op Cm +aCefnuvxIimqVEb
.Ek
.Bk -words
.Op Fl o Ar option_name
.Op Cm +o Ar option_name
.Ek
.Bk -words
.Op Ar argument ...
.Ek
.Sh DESCRIPTION
.Nm
is the standard command interpreter for the system.
The current version of
.Nm
is in the process of being changed to conform with the
.Tn POSIX
1003.2 and 1003.2a specifications for the shell.
This version has many
features which make it appear similar in some respects to the Korn shell,
but it is not a Korn shell clone (see
.Xr ksh 1 ) .
Only features designated by
.Tn POSIX ,
plus a few Berkeley extensions, are being incorporated into this shell.
This man page is not intended
to be a tutorial or a complete specification of the shell.
.Ss Overview
The shell is a command that reads lines from either a file or the
terminal, interprets them, and generally executes other commands.
It is the program that is running when a user logs into the system
(although a user can select a different shell with the
.Xr chsh 1
command).
The shell implements a language that has flow control
constructs, a macro facility that provides a variety of features in
addition to data storage, along with built in history and line editing
capabilities.
It incorporates many features to aid interactive use and
has the advantage that the interpretative language is common to both
interactive and non-interactive use (shell scripts).
That is, commands
can be typed directly to the running shell or can be put into a file and
the file can be executed directly by the shell.
.Ss Invocation
If no args are present and if the standard input of the shell
is connected to a terminal (or if the
.Fl i
flag is set),
and the
.Fl c
option is not present, the shell is considered an interactive shell.
An interactive shell generally prompts before each command and handles
programming and command errors differently (as described below).
When first starting,
the shell inspects argument 0, and if it begins with a dash
.Sq - ,
the shell is also considered
a login shell.
This is normally done automatically by the system
when the user first logs in.
A login shell first reads commands
from the files
.Pa /etc/profile
and
.Pa .profile
if they exist.
If the environment variable
.Ev ENV
is set on entry to an interactive shell, or is set in the
.Pa .profile
of a login shell, the shell next reads
commands from the file named in
.Ev ENV .
Therefore, a user should place commands that are to be executed only at
login time in the
.Pa .profile
file, and commands that are executed for every interactive shell inside the
.Ev ENV
file.
To set the
.Ev ENV
variable to some file, place the following line in your
.Pa .profile
of your home directory
.Pp
.Dl ENV=$HOME/.shinit; export ENV
.Pp
substituting for
.Dq .shinit
any filename you wish.
.Pp
If command line arguments besides the options have been specified, then
the shell treats the first argument as the name of a file from which to
read commands (a shell script), and the remaining arguments are set as the
positional parameters of the shell ($1, $2, etc).
Otherwise, the shell
reads commands from its standard input.
.Ss Argument List Processing
All of the single letter options that have a corresponding name can be
used as an argument to the
.Fl o
option.
The set
.Fl o
name is provided next to the single letter option in
the description below.
Specifying a dash
.Dq -
turns the option on, while using a plus
.Dq +
disables the option.
The following options can be set from the command line or
with the
.Ic set
builtin (described later).
.Bl -tag -width aaaallexportfoo -offset indent
.It Fl a Em allexport
Export all variables assigned to.
.It Fl c
Read commands from the
.Ar command_string
operand instead of from the standard input.
Special parameter 0 will be set from the
.Ar command_name
operand and the positional parameters ($1, $2, etc.)
set from the remaining argument operands.
.It Fl C Em noclobber
Don't overwrite existing files with
.Dq \*[Gt] .
.It Fl e Em errexit
If not interactive, exit immediately if any untested command fails.
The exit status of a command is considered to be
explicitly tested if the command is used to control an
.Ic if ,
.Ic elif ,
.Ic while ,
or
.Ic until ;
or if the command is the left hand operand of an
.Dq &&
or
.Dq ||
operator.
.It Fl f Em noglob
Disable pathname expansion.
.It Fl n Em noexec
If not interactive, read commands but do not execute them.
This is useful for checking the syntax of shell scripts.
.It Fl u Em nounset
Write a message to standard error when attempting to expand a variable
that is not set, and if the shell is not interactive, exit immediately.
.It Fl v Em verbose
The shell writes its input to standard error as it is read.
Useful for debugging.
.It Fl x Em xtrace
Write each command to standard error (preceded by a
.Sq +\  )
before it is executed.
Useful for debugging.
.It Fl I Em ignoreeof
Ignore EOF's from input when interactive.
.It Fl i Em interactive
Force the shell to behave interactively.
.It Fl l
Make dash act as if it had been invoked as a login shell.
.It Fl m Em monitor
Turn on job control (set automatically when interactive).
.It Fl s Em stdin
Read commands from standard input (set automatically if no file arguments
are present).
This option has no effect when set after the shell has
already started running (i.e. with
.Ic set ) .
.It Fl V Em vi
Enable the built-in
.Xr vi 1
command line editor (disables
.Fl E
if it has been set).
.It Fl E Em emacs
Enable the built-in
.Xr emacs 1
command line editor (disables
.Fl V
if it has been set).
.It Fl b Em notify
Enable asynchronous notification of background job completion.
(UNIMPLEMENTED for 4.4alpha)
.El
.Ss Lexical Structure
The shell reads input in terms of lines from a file and breaks it up into
words at whitespace (blanks and tabs), and at certain sequences of
characters that are special to the shell called
.Dq operators .
There are two types of operators: control operators and redirection
operators (their meaning is discussed later).
Following is a list of operators:
.Bl -ohang -offset indent
.It "Control operators:"
.Dl &  &&  \&(  \&)  \&;  ;; | || \*[Lt]newline\*[Gt]
.It "Redirection operators:"
.Dl \*[Lt]  \*[Gt]  \*[Gt]|  \*[Lt]\*[Lt]  \*[Gt]\*[Gt]  \*[Lt]&  \*[Gt]&  \*[Lt]\*[Lt]-  \*[Lt]\*[Gt]
.El
.Ss Quoting
Quoting is used to remove the special meaning of certain characters or
words to the shell, such as operators, whitespace, or keywords.
There are three types of quoting: matched single quotes,
matched double quotes, and backslash.
.Ss Backslash
A backslash preserves the literal meaning of the following
character, with the exception of
.Aq newline .
A backslash preceding a
.Aq newline
is treated as a line continuation.
.Ss Single Quotes
Enclosing characters in single quotes preserves the literal meaning of all
the characters (except single quotes, making it impossible to put
single-quotes in a single-quoted string).
.Ss Double Quotes

ChangeLog
2014-11-17  Stéphane Aulery <saulery@free.fr>

	* Correct typo in manual page.

2014-10-30  Herbert Xu <herbert@gondor.apana.org.au>

	* Catch variable length expansions on non-existant specials.

2014-10-28  Herbert Xu <herbert@gondor.apana.org.au>

	* Removed unnecessary pungetc on EOF from parser.
	* Simplify EOF/newline handling in list parser.

2014-10-27  Herbert Xu <herbert@gondor.apana.org.au>

	* Add printf support for format string a, A, and F.
	* Handle embedded NULs correctly in printf.
	* Replace open-coded flushall in preadbuffer.
	* Add likely tag in outmem.
	* Add ifdefs around MEM_OUT handling in outmem.
	* Remove unnecessary restoration of format string in printf.
	* Remove getintmax in printf.
	* Use error instead of warnx for fatal errors in printf.
	* Optimise handling of backslash octals in printf.
	* Simplify echo command.
	* Handle -- in dotcmd.

2014-10-13  Eric Blake <eblake@redhat.com>

	* cd: support drive letters on Cygwin.

2014-10-08  Herbert Xu <herbert@gondor.apana.org.au>

	* Split unquoted $@/$* correctly when IFS is set but empty.
	* Do not split quoted VSLENGTH and VSTRIM.
	* Optimise nulonly away and just use quoted as before.

2014-10-07  Herbert Xu <herbert@gondor.apana.org.au>

	* Use setvareq to set OPTIND initially.

2014-10-06  Herbert Xu <herbert@gondor.apana.org.au>

	* Exit without arguments in a trap should use status outside traps.
	* Do not allow break to break across function calls.
	* Move common skipcount logic into skiploop.
	* Allow return in loop conditional to set exit status.
	* Return without arguments in a trap should use status outside traps.

2014-10-03  Herbert Xu <herbert@gondor.apana.org.au>

	* Do not clobber exitstatus in evalcommand.

2014-10-02  Herbert Xu <herbert@gondor.apana.org.au>

	* Fix use-after-free in dotrap/evalstring.
	* Make sure evalskip is zero before running traps.
	* Set exitstatus in onint.

2014-09-29  Herbert Xu <herbert@gondor.apana.org.au>

	* Kill pgetc_macro.
	* Handle backslash newlines properly after dollar sign.
	* Add nlprompt/nlnoprompt helpers.

2014-09-28  Herbert Xu <herbert@gondor.apana.org.au>

	* Correctly handle test ! ! = !.

2014-09-26  Herbert Xu <herbert@gondor.apana.org.au>

	* Small optimisation of command -pv change.
	* Set command -p path to /usr/sbin:/usr/bin:/sbin:/bin.
	* Change "characters" for printf precision to "bytes".
	* Release 0.5.8.

2014-09-26  Jonathan Nieder <jrnieder@gmail.com>

	* Clarify "trap '' signals" syntax in manual page.

2014-09-26  Adam Buchbinder <adam.buchbinder@gmail.com>

	* Clarify that 0 and EXIT are both acceptable for exit traps.

2014-09-26  Harald van Dijk <harald@gigawatt.nl>

	* command: allow combining -p with -v.
	* getjob: Fix off-by-one error for multiple of four job numbers

2013-08-23  Herbert Xu <herbert@gondor.apana.org.au>

	* Propagate EXP_QPAT in subevalvar.
	* Initialise OPTIND after importing environment.
	* Fixed argument parsing crash in test.

2013-03-12  Peter Rosin <peda@lysator.liu.se>

	* Add newline when tracing in poplocalvars.

2013-01-10  Jérémie Courrèges-Anglas <jca+dash@wxcvbn.org>

	* Markup fixes in manual for mandoc 1.12.1.

2012-12-03  Harald van Dijk <harald@gigawatt.nl>

	* Use PRIdMAX instead of %j in printf.

2012-07-20  Kimo Rosenbaum <kimor79@yahoo.com>

	* Fix typo for wait in manual.

2012-06-09  Christoph Mathys <eraserix@gmail.com>

	* Add support for ulimit -r.

2012-03-11  Jim Meyering  <meyering@redhat.com>

	* Avoid overflow for very long variable name.
	$ perl -le 'print "v"x(2**31+1) ."=1"' | dash

2012-02-25  Herbert Xu <herbert@gondor.apana.org.au>

	* Sanitise environment variable names on entry.

2011-08-17  David S. Miller <davem@davemloft.net>

	* Allow building without LINEO support.
	* Add top-level autogen.sh.

2011-07-26  Harald van Dijk <harald@gigawatt.nl>

	* Avoid imaxdiv when only one of the results is wanted.

2010-07-09  maximilian attems <max@stro.at>

	* Fix klibc DEBUG compilation.

2011-07-09  Herbert Xu <herbert@gondor.apana.org.au>

	* Merge SKIPFUNC/SKIPFILE and only clear SKIPFUNC when leaving dotcmd.

2011-07-08  Herbert Xu <herbert@gondor.apana.org.au>

	* Release 0.5.7.

2011-07-07  Herbert Xu <herbert@gondor.apana.org.au>

	* Optimize dash -c "command" to avoid a fork.
	* Eliminate unnecessary promotion in echocmd.

2011-05-25  Jim Meyering  <meyering@redhat.com>

	* Avoid using undefined handler.

2011-05-23  Jim Meyering  <meyering@redhat.com>

	* Avoid gcc warning: variable 'oldstackp' set but not used.
	* Avoid clang warning about dead store to "size".

2011-05-22  Jonathan Nieder <jrnieder@gmail.com>

	* Make outc an inline function.


dash.ml
open Printf
open Ctypes
open Ctypes_types
open Foreign

type stackmark
              
let stackmark : stackmark structure typ = structure "stackmark"
let stackp = field stackmark "stackp" (ptr void)
let nxt = field stackmark "nxt" string
let size = field stackmark "stacknleft" PosixTypes.size_t
let () = seal stackmark

let init_stack () =
  let stack = make stackmark in (* ??? do we want to save this *)
  foreign "setstackmark" (ptr stackmark @-> returning void) (addr stack)
       
let init : unit -> unit = foreign "init" (void @-> returning void)

let initialize () =
  init ();
  init_stack ()
                                  
let setinputstring : string -> unit =
  foreign "setinputstring" (string @-> returning void)

let setinputtostdin () : unit =
  foreign "setinputfd" (int @-> int @-> returning void) 0 0 (* don't both pushing the file *)

let setinputfile (s : string) : unit =
  let _ = foreign "setinputfile" (string @-> int @-> returning int) s 0 in
  ()
          
(* first, we define the node type... *)
          
type node       
let node : node union typ = union "node"
let node_type = field node "type" int
(* but we don't seal it yet! *)

type nodelist
let nodelist : nodelist structure typ = structure "nodelist"       
let nodelist_next = field nodelist "next" (ptr nodelist)
let nodelist_n = field nodelist "n" (ptr node)
let () = seal nodelist
                       
type ncmd

let ncmd : ncmd structure typ = structure "ncmd"
let ncmd_type = field ncmd "type" int
let ncmd_linno = field ncmd "linno" int
let ncmd_assign = field ncmd "assign" (ptr node)
let ncmd_args = field ncmd "args" (ptr node)
let ncmd_redirect = field ncmd "redirect" (ptr node)
let () = seal ncmd

let node_ncmd = field node "ncmd" ncmd

type npipe

let npipe : npipe structure typ = structure "npipe"
let npipe_type = field npipe "type" int
let npipe_backgnd = field npipe "backgnd" int
let npipe_cmdlist = field npipe "cmdlist" (ptr nodelist)
let () = seal npipe

let node_npipe = field node "npipe" npipe
                           
type nredir

let nredir : nredir structure typ = structure "nredir"
let nredir_type = field nredir "type" int
let nredir_linno = field nredir "linno" int
let nredir_n = field nredir "n" (ptr node)
let nredir_redirect = field nredir "redirect" (ptr node)
let () = seal nredir

let node_nredir = field node "nredir" nredir

type nbinary

let nbinary : nbinary structure typ = structure "nbinary"
let nbinary_type = field nbinary "type" int
let nbinary_ch1 = field nbinary "ch1" (ptr node)
let nbinary_ch2 = field nbinary "ch2" (ptr node)
let () = seal nbinary

let node_nbinary = field node "nbinary" nbinary

type nif

let nif : nif structure typ = structure "nif"
let nif_type = field nif "type" int
let nif_test = field nif "test" (ptr node)
let nif_ifpart = field nif "ifpart" (ptr node)
let nif_elsepart = field nif "elsepart" (ptr node)
let () = seal nif

let node_nif = field node "nif" nif

type nfor

let nfor : nfor structure typ = structure "nfor"
let nfor_type = field nfor "type" int
let nfor_linno = field nfor "linno" int
let nfor_args = field nfor "args" (ptr node)
let nfor_body = field nfor "body" (ptr node)
let nfor_var = field nfor "var" string
let () = seal nfor

let node_nfor = field node "nfor" nfor

type ncase

let ncase : ncase structure typ = structure "ncase"
let ncase_type = field ncase "type" int
let ncase_linno = field ncase "linno" int
let ncase_expr = field ncase "expr" (ptr node)
let ncase_cases = field ncase "cases" (ptr node)
let () = seal ncase

let node_ncase = field node "ncase" ncase

type nclist

let nclist : nclist structure typ = structure "nclist"
let nclist_type = field nclist "type" int
let nclist_next = field nclist "next" (ptr node)
let nclist_pattern = field nclist "pattern" (ptr node)
let nclist_body = field nclist "body" (ptr node)
let () = seal nclist

let node_nclist = field node "nclist" nclist

type ndefun

let ndefun : ndefun structure typ = structure "ndefun"
let ndefun_type = field ndefun "type" int
let ndefun_linno = field ndefun "linno" int
let ndefun_text = field ndefun "text" string
let ndefun_body = field ndefun "body" (ptr node)
let () = seal ndefun

let node_ndefun = field node "ndefun" ndefun

type narg

let narg : narg structure typ = structure "narg"
let narg_type = field narg "type" int
let narg_next = field narg "next" (ptr node)
let narg_text = field narg "text" string
let narg_backquote = field narg "backquote" (ptr nodelist)
let () = seal narg

let node_narg = field node "narg" narg

type nfile

let nfile : nfile structure typ = structure "nfile"
let nfile_type = field nfile "type" int
let nfile_next = field nfile "next" (ptr node)
let nfile_fd = field nfile "fd" int
let nfile_fname = field nfile "fname" (ptr node)
let nfile_expfname = field nfile "expfname" string
let () = seal nfile

let node_nfile = field node "nfile" nfile

type ndup

let ndup : ndup structure typ = structure "ndup"
let ndup_type = field ndup "type" int
let ndup_next = field ndup "next" (ptr node)
let ndup_fd = field ndup "fd" int
let ndup_dupfd = field ndup "dupfd" int
let ndup_vname = field ndup "vname" (ptr node)
let () = seal ndup

let node_ndup = field node "ndup" ndup

type nhere

let nhere : nhere structure typ = structure "nhere"
let nhere_type = field nhere "type" int
let nhere_next = field nhere "next" (ptr node)
let nhere_fd = field nhere "fd" int
let nhere_doc = field nhere "doc" (ptr node)
let () = seal nhere

let node_nhere = field node "nhere" nhere

type nnot

let nnot : nnot structure typ = structure "nnot"
let nnot_type = field nnot "type" int
let nnot_com = field nnot "com" (ptr node)
let () = seal nnot

let node_nnot = field node "nnot" nnot
let () = seal node
              
let parsecmd : int -> node union ptr =
  foreign "parsecmd" (int @-> returning (ptr node))
          
let parse s =
  setinputstring s; (* TODO set stack mark? *)
  parsecmd 0

let neof : node union ptr = foreign_value "tokpushback" node

let addrof p = raw_address_of_ptr (to_voidp p)

let eqptr p1 p2 = addrof p1 = addrof p2
                                  
let nullptr (p : 'a ptr) = addrof p = Nativeint.zero

let parse_next () =
  let n = parsecmd 0 in
  if eqptr n neof
  then `Done
  else if nullptr n
  then `Null (* comment or blank line ... *)
  else `Parsed n

let rec parse_all () : (node union ptr) list =
  match parse_next () with
  | `Done -> []
  | `Null -> parse_all ()
  | `Parsed n -> n::parse_all ()
            
let (@->) (s : ('b, 'c) structured ptr) (f : ('a, ('b, 'c) structured) field) =
  getf (!@ s) f

let rec arglist (n : narg structure) : (narg structure) list =
  let next = getf n narg_next in
  if nullptr next
  then [n] 
  else
    (assert (next @-> node_type = 15);
     n::arglist (next @-> node_narg))

let rec nodelist (n : nodelist structure ptr) : (node union ptr) list =
  if nullptr n
  then []
  else (n @-> nodelist_n)::nodelist (n @-> nodelist_next)
                  
let rec redirlist (n : node union ptr) =
  if nullptr n
  then []
  else
    let h = match n @-> node_type with
      (* NTO *)
      | 16 -> `File (1,">",n @-> node_nfile)
      (* NCLOBBER *)
      | 17 -> `File (1,">|",n @-> node_nfile)
      (* NFROM *)
      | 18 -> `File (0,"<",n @-> node_nfile)
      (* NFROMTO *)
      | 19 -> `File (0,"<>",n @-> node_nfile)
      (* NAPPEND *)
      | 20 -> `File (1,">>",n @-> node_nfile)
      (* NTOFD *)      
      | 21 -> `Dup (1,">&",n @-> node_ndup)
      (* NFROMFD *)              
      | 22 -> `Dup (0,"<&",n @-> node_ndup)
      (* NHERE quoted heredoc---no expansion)*)
      | 23 -> `Here (0,"<<",false,n @-> node_nhere)
      (* NXHERE unquoted heredoc (param/command/arith expansion) *)
      | 24 -> `Here (0,"<<",true,n @-> node_nhere)
      | nt -> failwith ("unexpected node_type in redirlist: " ^ string_of_int nt)
    in
    h :: redirlist (getf (n @-> node_nfile) nfile_next)

let rec caselist (n : node union ptr) =
  if nullptr n
  then []
  else    
    let n = n @-> node_nclist in
    assert (getf n nclist_type = 13); (* NCLIST *)
    (getf n nclist_pattern, getf n nclist_body)::caselist (getf n nclist_next)
                   
let explode s =
  let rec exp i l =
    if i < 0 then l else exp (i - 1) (s.[i] :: l) in
  exp (String.length s - 1) []

let implode l =
  let s = Bytes.create (List.length l) in
  let rec imp i l =
    match l with
    | []  -> ()
    | (c::l) -> (Bytes.set s i c; imp (i+1) l)
  in
  imp 0 l;
  Bytes.unsafe_to_string s
                   
let rec intercalate p ss =
  match ss with
  | [] -> ""
  | [s] -> s
  | s::ss -> s ^ p ^ intercalate p ss          

let lines = Str.split (Str.regexp "[\n\r]+")

let rec fresh_marker ls s =
  if List.mem s ls
  then fresh_marker ls (s ^ (String.sub s (String.length s - 1) 1))
  else s
                      
let rec split_at p xs =
  match xs with
  | [] -> ([],[])
  | x::xs ->
     if p x
     then ([],x::xs)
     else let (xs,ys) = split_at p xs in
          (x::xs, ys)

let string_of_vs = function
  | 0x1 -> (* VSNORMAL ${var} *) []
  | 0x2 -> (* VSMINUS ${var-text} *) ['-']
  | 0x3 -> (* VSPLUS ${var+text} *) ['+']
  | 0x4 -> (* VSQUESTION ${var?message} *) ['?']
  | 0x5 -> (* VSASSIGN ${var=text} *) ['=']
  | 0x6 -> (* VSTRIMRIGHT ${var%pattern} *) ['%']
  | 0x7 -> (* VSTRIMRIGHTMAX ${var%%pattern} *) ['%';'%']
  | 0x8 -> (* VSTRIMLEFT ${var#pattern} *) ['#']
  | 0x9 -> (* VSTRIMLEFTMAX ${var##pattern} *) ['#';'#']
  | vs -> failwith ("Unknown VSTYPE: " ^ string_of_int vs)
                   
let braces s = "{ " ^ s ^ " ; }"
let parens s = "( " ^ s ^ " )"
                  
let rec show (n : node union ptr) : string =
  match (n @-> node_type) with
  (* NCMD *)
  | 0  ->
     let n = n @-> node_ncmd in
     let raw_cmd = intercalate " " (List.map sharg (arglist (getf n ncmd_args @-> node_narg))) in
     let vars = if nullptr (getf n ncmd_assign) then "" else intercalate " " (List.map sharg (arglist (getf n ncmd_assign @-> node_narg))) ^ " " in
     vars ^ raw_cmd ^ shredir (getf n ncmd_redirect)
  (* NPIPE *)
  | 1  ->
     let n = n @-> node_npipe in
     let cmds = nodelist (getf n npipe_cmdlist) in
     intercalate " | " (List.map show cmds) ^ if (getf n npipe_backgnd) = 0 then "" else " &"
  (* NREDIR *)
  | 2  -> shnredir braces n 
  (* NBACKGND *)
  | 3  -> shnredir braces n ^ " &"
  (* NSUBSHELL *)
  | 4  -> shnredir parens n
  (* NAND *)
  | 5  -> shbinary "&&" (n @-> node_nbinary)
  (* NOR *)
  | 6  -> shbinary "||" (n @-> node_nbinary)
  (* NSEMI *)
  | 7  -> shbinary ";" (n @-> node_nbinary)
  (* NIF *)
  | 8  -> shif (n @-> node_nif)
  (* NWHILE *)
  | 9  ->
     let n = n @-> node_nbinary in
     "while " ^ show (getf n nbinary_ch1) ^ "; do " ^ show (getf n nbinary_ch2) ^ "; done"
  (* NUNTIL *)
  | 10 ->
     let n = n @-> node_nbinary in
     "until " ^ show (getf n nbinary_ch1) ^ "; do " ^ show (getf n nbinary_ch2) ^ "; done"
  (* NFOR *)
  | 11 ->
     let n = n @-> node_nfor in
     "for " ^ (getf n nfor_var) ^ " in " ^ sharg (getf n nfor_args @-> node_narg) ^ "; do " ^ show (getf n nfor_body) ^ "; done"
  (* NCASE *)
  | 12 ->
     let n = n @-> node_ncase in
     "case " ^ sharg (getf n ncase_expr @-> node_narg) ^ " in " ^ shclist (getf n ncase_cases) ^ " esac"
  (* NDEFUN *)
  | 14 ->
     let n = n @-> node_ndefun in
     (getf n ndefun_text) ^ "() " ^ braces (show (getf n ndefun_body))
  (* NARG *)
  | 15 -> failwith "Didn't expect narg at the top-level"
  (* NNOT *)
  | 25 -> "! { " ^ show (getf (n @-> node_nnot) nnot_com) ^ " }"
  | nt -> failwith ("unexpected node_type " ^ string_of_int nt)

and shbinary (op : string) (n : nbinary structure) : string =
  show (getf n nbinary_ch1) ^ " " ^ op ^ " " ^ show (getf n nbinary_ch2)

and shnredir parenthesize n =
  let nr = n @-> node_nredir in
  parenthesize (show (getf nr nredir_n)) ^ shredir (getf nr nredir_redirect)

and shif n =
  "if " ^ show (getf n nif_test) ^
  "; then " ^ show (getf n nif_ifpart) ^
  (let else_part = getf n nif_elsepart in
   if nullptr else_part
   then "; fi"
   else if (else_part @-> node_type = 8)
   then "; el" ^ shif (else_part @-> node_nif)
   else "; else " ^ show else_part ^ "; fi")

and shclist clist = intercalate " " (List.map shcase (caselist clist)) (* handles NCLIST = 13 *)
    
and shcase (pat,body) =
  assert (pat @-> node_type = 15);
  sharg (pat @-> node_narg) ^ ") " ^ show body ^ ";;"
    
and shredir (n : node union ptr) : string =
  let redirs = redirlist n in
  if redirs = []
  then ""
  else " " ^ intercalate " " (List.map show_redir redirs)
and show_redir n : string =
  match n with
  | `File (src,sym,f) -> show_redir_src (getf f nfile_fd) src ^ sym ^ sharg ((getf f nfile_fname) @-> node_narg)
  | `Dup (src,sym,d) -> show_redir_src (getf d ndup_fd) src ^ sym ^ string_of_int (getf d ndup_dupfd)
  | `Here (src,sym,exp,h) ->
     let heredoc = sharg ((getf h nhere_doc) @-> node_narg) in
     let marker = fresh_marker (lines heredoc) "EOF" in
     show_redir_src (getf h nhere_fd) src ^ sym ^ (if exp then marker else "'" ^ marker ^ "'") ^ "\n" ^ heredoc ^ marker
and show_redir_src actual expected =
  if actual = expected
  then ""
  else string_of_int actual
                                                    
and sharg (n : narg structure) : string =
  let str,s',bqlist,stack = show_arg (explode (getf n narg_text)) (getf n narg_backquote) [] in
  (* we should have used up the string and have no backquotes left in our list *)
  assert (s' = []);
  assert (nullptr bqlist);
  assert (stack = []);
  str    
and show_arg (s : char list) (bqlist : nodelist structure ptr) stack =
  (* we have to look at the string and interpret control characters... *)
  match s,stack with
  | [],[] -> "",[],bqlist,[]
  | [],`CTLVar::stack' -> failwith "End of string before CTLENDVAR"
  | [],`CTLAri::stack' -> failwith "End of string before CTLENDARI"
  | [],`CTLQuo::stack' -> failwith "End of string before CTLQUOTEMARK"
  (* CTLESC *)
  | '\129'::c::s',_ -> 
     let str,s'',bqlist',stack' = show_arg s' bqlist stack in
     let c' = match c with
      | '\'' -> "\\'"
      | '\"' -> "\\\""
      | _ -> Char.escaped c
     in
     c' ^ str,s'',bqlist',stack'
  (* CTLVAR *)
  | '\130'::t::s',_ -> 
     let v,s'',bqlist',stack' = show_var (int_of_char t) s' bqlist stack in
     assert (stack = stack');
     let str,s''',bqlist'',stack'' = show_arg s'' bqlist' stack' in
     "${" ^ v ^ "}" ^ str, s''', bqlist'', stack''
  (* CTLENDVAR *)
  | '\131'::s',`CTLVar::stack' -> "",[],bqlist,stack' (* s' gets handled by CTLVAR *)
  | '\131'::s',`CTLAri::stack' -> failwith "Saw CTLENDVAR before CTLENDARI"
  | '\131'::s',`CTLQuo::stack' -> failwith "Saw CTLENDVAR before CTLQUOTEMARK"
  | '\131'::s',[] -> failwith "Saw CTLENDVAR outside of CTLVAR"
  (* CTLBACKQ *)
  | '\132'::s',_ ->
     if nullptr bqlist
     then failwith "Saw CTLBACKQ but bqlist was null"
     else
       let n = bqlist @-> nodelist_n in
       (* MMG: !!! dash has a bug in its sharg function... it doesn't advance the list! *)
       let bqlist' = bqlist @-> nodelist_next in
       let str,s'',bqlist'',stack' = show_arg s' bqlist' stack in
       "$(" ^ show n ^ ")" ^ str,s'',bqlist'',stack'
  (* CTLARI *)
  | '\134'::s',_ ->
     let ari,s'',bqlist',stack' = show_arg s' bqlist (`CTLAri::stack) in
     assert (stack = stack');
     let str,s''',bqlist'',stack'' = show_arg s'' bqlist' stack' in
     "$((" ^ ari ^ "))" ^ str, s''', bqlist'', stack''
  (* CTLENDARI *)
  | '\135'::s',`CTLAri::stack' -> "",s',bqlist,stack'
  | '\135'::s',`CTLVar::stack' -> failwith "Saw CTLENDARI before CTLENDVAR"
  | '\135'::s',`CTLQuo::stack' -> failwith "Saw CTLENDARI before CTLQUOTEMARK"
  | '\135'::s',[] -> failwith "Saw CTLENDARI outside of CTLARI"
  (* CTLQUOTEMARK *)
  | '\136'::s',[`CTLQuo] -> "",s',bqlist,[]
  | '\136'::s',_ ->
     let quoted,s'',bqlist',stack' = show_arg  s' bqlist [`CTLQuo] in
     assert (stack' = []);
     let str,s''',bqlist'',stack'' = show_arg s'' bqlist' stack in
     "\"" ^ quoted ^ "\"" ^ str, s''', bqlist'', stack''
  (* ordinary character *)
  | c::s',_ -> 
     let str,s',bqlist',stack' = show_arg s' bqlist stack in
     let c' = match c with
      | '\'' -> "\\'"
      | '\"' -> "\\\""
      | _ -> String.make 1 c
     in
     c' ^ str,s',bqlist',stack'
and show_var (t : int) (s : char list) (bqlist : nodelist structure ptr) stack =
  let var_name,s' = split_at (fun c -> c = '=') s in
  (* mask out VSNUL, check VSTYPE *)
  match t land 0x0f, s' with
  (* VSNORMAL and VSLENGTH get special treatment

     neither ever gets VSNUL
     VSNORMAL is terminated just with the =, without a CTLENDVAR *)
  (* VSNORMAL *)
  | 0x1,'='::s'' -> implode var_name, s'', bqlist, stack
  (* VSLENGTH *)
  | 0xa,'='::'\131'::s'' -> implode (['#'] @ var_name), s'', bqlist, stack
  | 0x1,c::_ | 0xa,c::_ -> failwith ("Missing CTLENDVAR for VSNORMAL/VSLENGTH, found " ^ Char.escaped c)
  (* every other VSTYPE takes mods before CTLENDVAR *)
  | vstype,'='::s' ->
     (* check VSNUL *)
     let vsnul = if t land 0x10 = 1 then [] else [':'] in
     let mods,s'',bqlist',stack' = show_arg s' bqlist (`CTLVar::stack) in
     implode (var_name @ vsnul @ string_of_vs vstype) ^ mods, s'', bqlist', stack'
  | _,c::s' -> failwith ("Expected '=' terminating variable name, found " ^ Char.escaped c)
  | _,[] -> failwith "Expected '=' terminating variable name, found EOF"

